### PR TITLE
Make IDataLoaderContextAccessor.Context non-null

### PR DESCRIPTION
**Pull request by ccurrens**
As discussed in #3777, this PR changes the null annotations for `IDataLoaderContextAccessor.Context` to be non-null. This supports the nullability contract when the property is used correctly (e.g., inside of a document execution) and when the `DataLoaderDocumentListener` is registered

### CODE REVIEW

File: src/GraphQL.DataLoader/DataLoaderContextAccessor.cs
Comment (ccurrens) [inline src/GraphQL.DataLoader/DataLoaderContextAccessor.cs:12]: I'm a little unsure about this. On the one hand, using `!` here will result in the least amount of code when the data loader feature is used correctly. But when used incorrectly, it returns null rather than having a helpful exception message. I have a hard time penalizing the correct case for this, though.

An alternative would be to use a throw helper to throw when it is null. The property could still be inlined, but there's still a conditional check...

Comment (ccurrens) [reply src/GraphQL.DataLoader/DataLoaderContextAccessor.cs:12]: Ah, another option I just thought of: `Debug.Assert`. That will help developers ensure it is setup correctly while not penalizing production use.

Comment (Shane32) [reply src/GraphQL.DataLoader/DataLoaderContextAccessor.cs]: Debug.Assert are for internal testing; the compiler removes these from the compiled nuget packages. So developers won’t get any benefit from Debug.Assert.

Comment (ccurrens) [reply src/GraphQL.DataLoader/DataLoaderContextAccessor.cs]: 🤦‍♂️ of course. I'll remove that. Do you think I should add a throw helper or leave it without one?

Comment (Shane32) [reply src/GraphQL.DataLoader/DataLoaderContextAccessor.cs]: I probably would. I can’t think of a reason you’d need to know the value was null. However I’d consider it a breaking change and push the change to v8

Comment (Shane32) [reply src/GraphQL.DataLoader/DataLoaderContextAccessor.cs]: Adding the throw helper assists the developer in knowing how to fix the issue which is a great benefit.

Comment (ccurrens) [reply src/GraphQL.DataLoader/DataLoaderContextAccessor.cs]: Great, I'll change the base to the develop branch and add the throw helper.

File: src/GraphQL.DataLoader/DataLoaderContextAccessor.cs
Comment (Shane32) [inline src/GraphQL.DataLoader/DataLoaderContextAccessor.cs:18]: A few comments here:
- By default, neither GraphQL.NET nor GraphQL.NET Server pull all `IDocumentExecutionListener` instances from DI, so registering the listener in the IoC container is not enough to fix the problem.  While it is possible with custom code to do so, the current suggested mechanism is to use `.AddDataLoaders()` to install the data loader document listener.  This will register `DataLoaderDocumentListener` in the DI container and install a `IConfigureExecution` instance which is set up to pull the listener from DI and add it to `ExecutionOptions.Listeners`.
- By convention, we end all exception messages with a period.

```suggestion
        => throw new InvalidOperationException("DataLoaderContext is null. Ensure that you have called IGraphQLBuilder.AddDataLoader() or otherwise have added an instance of DataLoaderDocumentListener to ExecutionOptions.Listeners.");
```

File: src/GraphQL.DataLoader/IDataLoaderContextAccessor.cs
Comment (Shane32) [inline src/GraphQL.DataLoader/IDataLoaderContextAccessor.cs:12]: 👍

**Comment by ccurrens** [comment]
Whoops, looks like there are a few test failures. Didn't run it after adding the thrown exception. I'll take a look

File: src/GraphQL.DataLoader/DataLoaderDocumentListener.cs
Comment (ccurrens) [inline src/GraphQL.DataLoader/DataLoaderDocumentListener.cs:29]: The tests were failing because this did a null check before setting to a new instance. That no longer works now that we throw if it's `null`.

The obvious fix was to add a function named `Initialize` that could bypass the getter, accessing the `AsyncLocal` directly and setting it to a new instance if it was `null`. I then also decided to remove the setter altogether and introduce a function to reset the context back to `null`.

I'm not sure the setter makes sense anymore since you can't do a conditional set based on whether or not it is `null`. This is why I went ahead and removed it and replace it with a `Reset()` function to set it back to `null`.

As I write this, I realize it's a larger breaking change because consumers can no longer set the context to an arbitrary value. So maybe a) the setter should remain or b) `Reset` should take in a `DataLoaderContext?` parameter so allow consumers to set it to whatever they want.

Comment (Shane32) [reply src/GraphQL.DataLoader/DataLoaderDocumentListener.cs:29]: Maybe we should reconsider simply changing the NRT attributes. 

Also consider that this interface is modeled after `IHttpContextAccessor` and we should review/consider their implementation. 

https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.ihttpcontextaccessor.httpcontext?view=aspnetcore-7.0#microsoft-aspnetcore-http-ihttpcontextaccessor-httpcontext

Comment (Shane32) [reply src/GraphQL.DataLoader/DataLoaderDocumentListener.cs:29]: Another option is to leave existing property as is and introduce a new property like RequiredContext (similar in vein to GetRequiredService). Sorta wish it was an extension method though

Comment (ccurrens) [reply src/GraphQL.DataLoader/DataLoaderDocumentListener.cs:29]: Regarding it being modeled against `IHttpContextAccessor`, I think a good question to ask is what are valid use cases for it. In one [documentation example](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/use-http-context?view=aspnetcore-7.0#httpcontext-isnt-thread-safe) the `IHttpContextAccessor` is used in a `HttpClient` delegating handler, where depending on when the `HttpClient` is called, could result in being outside of the request flow and having the `HttpContext` nullable.

Is it considered valid usage of the `DataLoaderContext` for it to be accessed outside of document execution? If it is, the property should stay nullable and something like `RequiredContext`/`GetRequiredContext` could be introduced. If it's not valid to use it outside of document execution, then I think it's better to have it modeled against the valid use case (non-null).

----------------

Another interesting discussion on nullability is [here](https://github.com/dotnet/aspnetcore/pull/22928#discussion_r440513476), when `ISession` is changed to be non-null, even though technically custom implementations of `HttpContext` could cause it to return a non-null instance. I like the comment that there needs to be pragmatism applied to nullability attributes and focus on the 99% use case, and I think their [implementation in `DefaultHttpContext`](https://github.com/dotnet/aspnetcore/blob/744cb017a7d7d14a5e25076b9a9a8d3e4702f5d8/src/Http/Http/src/DefaultHttpContext.cs#L198) is similar to what I'm attempting here in the previous commit.

So another possible option is to revert the last commit that adds `Reset` and `Initialize`, which effectively disallows the `??=` operator. But this is not really different from what was done with `DefaultHttpContext.Session`. This makes the assumption that changing the way `DataLoaderDocumentListener` sets the context is safe:

```csharp
// before
_accessor.Context ??= new();

// after
_accessor.Context = new();
```
It seems like it's safe to skip the null check and always set it (I assume each `IDocumentExecutionListener` is only executed once). If a custom listener sets their own, it would require that they ensure `DataLoaderDocumentListener` is not added.

Comment (Shane32) [reply src/GraphQL.DataLoader/DataLoaderDocumentListener.cs:29]: Well, to think about it a bit, having `IHttpContextAccessor` return a nullable context allows code to operate differently based on whether it was null, and is useful for logging and similar scenarios.  But pulling the `IDataLoaderContext` is really only useful if you're going to use the data loader, which implies returning an `IDataLoaderResult<T>` from the method.  Sure it's theoretically possible to detect that the context is null and immediately retrieve the values, but you'd still need to wrap the values in an `IDataLoaderResult<T>`.  As such, the only practical use would be for tests, and it's much simpler to mock the data loader context in that scenario.

So I'd have to say that it's not really 'valid usage' to use `IDataLoaderContextAccessor` without the listener in use.

Agree?

Comment (Shane32) [reply src/GraphQL.DataLoader/DataLoaderDocumentListener.cs:29]: Let's go with your suggestion above, reverting the last commit and changing this:

```cs
// before
_accessor.Context ??= new();

// after
_accessor.Context = new();
```

If someone wants to be able to read `null` values from the context, they can write their own implementation of `DataLoaderContextAccessor` which allows null values or provides `Initialize` and `Reset` methods or similar.

Comment (Shane32) [reply src/GraphQL.DataLoader/DataLoaderDocumentListener.cs:29]: (Which is how the PR is now.)

**Comment by Shane32** [comment]
You'll need to run the tests local so that the ApiApprovals test data is updated

File: src/GraphQL.DataLoader/DataLoaderDocumentListener.cs
Comment (ccurrens) [reply src/GraphQL.DataLoader/DataLoaderDocumentListener.cs:29]: > So I'd have to say that it's not really 'valid usage' to use IDataLoaderContextAccessor without the listener in use.

I very much agree. I was thinking about it yesterday and came to the same conclusion about the data loader result.

**Review by Shane32 — APPROVED** [review]

**Comment by Shane32** [comment]
Looks good to me.  I'll leave open for a few days in case any else has any comments before merging.

cc: @gao-artur

**Comment by gao-artur** [comment]
Looks good 👍

**Pull request merged**
